drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+		volatile u32_t *intr;
+		volatile u32_t *intr_clr;
+		volatile u32_t *intr_status;
+		volatile u32_t *intr_raw;
+		volatile u32_t *status;
+		volatile u32_t *cmd_base;
+		volatile u32_t *timeout;
+		volatile u32_t *scl_low_period;
+		volatile u32_t *scl_high_period;
+		volatile u32_t *scl_start_h_time;
+		volatile u32_t *scl_rstart_s_time;
+		volatile u32_t *scl_stop_h_time;
+		volatile u32_t *scl_stop_s_time;
+		volatile u32_t *sda_hold_time;
+		volatile u32_t *sda_sample_time;
+	} regs;
+
+	const struct {
+		int sda_out;
+		int sda_in;
+		int scl_out;
+		int scl_in;
+	} sig;
+
+	const struct {
+		int scl;
+		int sda;
+		bool scl_pull_up;
+		bool sda_pull_up;
+		bool default_level;
+	} pins;
+
+	const struct {
+		int clk;
+		int rst;
+	} enable_bits;
+
+	const struct {
+		bool tx_lsb_first;
+		bool rx_lsb_first;
+	} mode;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	const union dev_config default_config;
+};
+
+static void i2c_esp32_set_device_enabled(const struct i2c_esp32_config *config,
+					 bool setting)
+{
+	volatile u32_t *periph_clk_reg = (u32_t *)DPORT_PERIP_CLK_EN_REG;
+	volatile u32_t *periph_rst_reg = (u32_t *)DPORT_PERIP_RST_EN_REG;
+
+	if (setting) {
+		*periph_clk_reg |= config->enable_bits.clk;
+		*periph_rst_reg &= ~config->enable_bits.rst;
+	} else {
+		*periph_clk_reg &= ~config->enable_bits.clk;
+		*periph_rst_reg |= config->enable_bits.rst;
+	}
+}
+
+static const char *i2c_esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+
+static int i2c_esp32_configure_pins(int pin, bool pull_up, bool default_level,
+				    int matrix_out, int matrix_in)
+{
+	const char *device_name = i2c_esp32_get_gpio_for_pin(pin);
+	struct device *gpio;
+
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	gpio_pin_configure(gpio, pin,
+			   GPIO_DIR_OUT | GPIO_DS_DISCONNECT_LOW |
+			   pull_up ? GPIO_PUD_PULL_UP : GPIO_PUD_NORMAL);
+	gpio_pin_write(gpio, pin, default_level);
+
+	gpio_matrix_out(pin, matrix_out, 0, 0);
+	gpio_matrix_in(pin, matrix_in, 0);
+
+	return 0;
+}
+
+static void i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					  u32_t freq_hz)
+{
+	const u32_t period = (APB_CLK_FREQ / freq_hz) / 2;
+
+	*config->regs.scl_low_period = period << I2C_SCL_LOW_PERIOD_S;
+	*config->regs.scl_high_period = period << I2C_SCL_HIGH_PERIOD_S;
+
+	*config->regs.scl_start_h_time =
+		(period / 2) << I2C_SCL_START_HOLD_TIME_S;
+	*config->regs.scl_rstart_s_time =
+		(period / 2) << I2C_SCL_RSTART_SETUP_TIME_S;
+	*config->regs.scl_stop_h_time =
+		(period / 2) << I2C_SCL_STOP_HOLD_TIME_S;
+	*config->regs.scl_stop_s_time =
+		(period / 2) << I2C_SCL_STOP_SETUP_TIME_S;
```

Comment:
If those are shift/exponent values and not linear multipliers, should probably have more than just a _S appended to cue the reader.

Reply:
These are defined by esp-idf.